pkg/backend/httpstate/client/client.go

```diff
@@ -607,16 +607,21 @@ func (pc *Client) PublishPolicyPack(ctx context.Context, orgName string,
 	}
 
 	//
-	// Step 2: Upload the compressed PolicyPack directory to the presigned S3 URL. The PolicyPack is
-	// now published.
+	// Step 2: Upload the compressed PolicyPack directory to the pre-signed object storage service URL.
+	// The PolicyPack is now published.
 	//
 
-	putS3Req, err := http.NewRequest(http.MethodPut, resp.UploadURI, dirArchive)
+	putReq, err := http.NewRequest(http.MethodPut, resp.UploadURI, dirArchive)
 	if err != nil {
 		return "", errors.Wrapf(err, "Failed to upload compressed PolicyPack")
 	}
 
-	_, err = http.DefaultClient.Do(putS3Req)
+	// Add a custom header for Azure Storage so that if the pre-signed URL is for their
+	// storage service, the request doesn't fail due to it being required.
+	// See x-ms-blob-type on https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob
+	putReq.Header.Add("x-ms-blob-type", "BlockBlob")
```

Comment:
I didn't add any conditional checks to include the header because storage accounts might have a custom domain and therefore the detection may not be 100% accurate. If you have other suggestions, please let me know.

Reply:
Maybes it's too much - but would it make sense for the service to return back required headers to include as part of the CreatePolicyPackResponse above?  Then this wouldn't need to be hard coded - and could be provided by the service which knows what kind of storage backend is being used?

Also - does this break S3 requests?  I thought the headers with part of the presigned signature and couldn't be modified like this?

Reply:
That's a good idea to include required headers in the response. I am wondering if we could simply return the backend object storage scheme? That way, the only header we would add is the custom header for Azure Storage. It would be sort of explicit.

About S3 requests and pre-signed URL restrictions: I believe if we were to set one of the [supported headers](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/s3-example-presigned-urls.html) ([gocloud ref](https://github.com/google/go-cloud/blob/master/blob/blob.go#L1059)), then, only the value that was given at the time of URL generation will be allowed. I will double-check that, though.

Reply:
>  I am wondering if we could simply return the backend object storage scheme? That way, the only header we would add is the custom header for Azure Storage. It would be sort of explicit.

Curious what you have in mind here?  You mean the client would key off of the scheme to decide what headers to include?  Still feels a little oddly coupled, but definitely workable if that's sufficiently easier than leaving all the logic about what URLs and headers to POST to in the service.

Reply:
> You mean the client would key off of the scheme to decide what headers to include?

Yes, this is what I am thinking.

> Still feels a little oddly coupled, but definitely workable if that's sufficiently easier than leaving all the logic about what URLs and headers to POST to in the service.

Yep agreed. Maybe this is what you were originally suggesting too, but we could return the headers with both the name **and** the value that need to be set by the client. Otherwise, the CLI will need to know what value to set anyway for a header and so we don't gain much. But it seems like a bad idea to include something in the response that the CLI implicitly trusts and round-trips to the signed URL. Rather I am fine with being explicit on the client-side about what header it will include based on the scheme. What do you think? The alternative is to leave this as-is for now, i.e. include this header for all signed URLs since it is a custom header.

Reply:
@leezen and I talked about the options here. We are going to go the route of having the service return the headers that the CLI should add in order to use the pre-signed URL. It will be in favor of us not wanting to couple the CLI with the storage mechanism. I'll update this PR once I have the service changes ready.